mpi4py: Regression with ch4:ofi

Looks like this regression was introduced in https://github.com/pmodels/mpich/pull/7503.
The fix in https://github.com/pmodels/mpich/pull/7512 does not resolve the mpi4py test failure.

```
testMessageProbeIProbe (test_util_pkl5.TestMPIWorld.testMessageProbeIProbe) ... 
Assertion failed in file src/mpid/ch4/netmod/ofi/ofi_init.c at line 884: 0
...
/usr/local/lib/libmpi_abi.so.0(MPI_Comm_dup+0x61) [0x7fc630d79cc7]
Abort(1) on node 1: Internal error
...
/usr/local/lib/libmpi_abi.so.0(MPI_Comm_dup+0x61) [0x7fe530779cc7]
Abort(1) on node 0: Internal error
Error: Process completed with exit code 1.
```
full logs: https://github.com/mpi4py/mpi4py-testing/actions/runs/16534706230/job/46767010420

Not sure whether this is a leak of CS, or some recursive invocation. For what is worth, my test is not calling `MPI_Comm_dup` within any MPI callback.

Note: I originally reported this problem in this comment: https://github.com/pmodels/mpich/pull/7503#issuecomment-3083855490, however out attention got diverted to error handling and CS enter/exit.

PS: I'll try later to provide a minimal command line reproducer using mpi4py's test suite. For the time being, running `mpi4py -n 2 python test/test_util_pkl5.py` should be hopefully enough to reproduce.

--- a/src/mpid/ch4/netmod/ofi/ofi_comm.c
+++ b/src/mpid/ch4/netmod/ofi/ofi_comm.c
@@ -200,17 +200,8 @@ int MPIDI_OFI_mpi_comm_free_hook(MPIR_Comm * comm)
 
     MPL_free(MPIDI_OFI_COMM(comm).pref_nic);
 
-    if (strcmp("sockets", MPIDI_OFI_global.prov_use[0]->fabric_attr->prov_name) == 0) {
-        /* sockets provider need flush any last lightweight send. */
-        mpi_errno = MPIDI_OFI_flush_send_queue();
-        MPIR_ERR_CHECK(mpi_errno);
-    }
-
-  fn_exit:
     MPIR_FUNC_EXIT;
     return mpi_errno;
-  fn_fail:
-    goto fn_exit;
 }
 
 int MPIDI_OFI_comm_set_hints(MPIR_Comm * comm, MPIR_Info * info)

--- a/src/mpid/ch4/netmod/ofi/ofi_impl.h
+++ b/src/mpid/ch4/netmod/ofi/ofi_impl.h
@@ -60,7 +60,6 @@ ATTRIBUTE((unused));
 
 int MPIDI_OFI_progress_uninlined(int vci);
 int MPIDI_OFI_handle_cq_error(int vci, int nic, ssize_t ret);
-int MPIDI_OFI_flush_send_queue(void);
 
 /*
  * Helper routines and macros for request completion

--- a/src/mpid/ch4/netmod/ofi/ofi_init.c
+++ b/src/mpid/ch4/netmod/ofi/ofi_init.c
@@ -800,8 +800,7 @@ static int flush_recv(int src, int nic, int vci, MPIDI_OFI_dynamic_process_reque
     goto fn_exit;
 }
 
-/* flush send_queue to ensure injected messages are out. Needed by some providers e.g. socket */
-int MPIDI_OFI_flush_send_queue(void)
+static int flush_send_queue(void)
 {
     int mpi_errno = MPI_SUCCESS;
 
@@ -811,11 +810,7 @@ int MPIDI_OFI_flush_send_queue(void)
 
     /* Apparently by sending self messages can flush the send queue */
     int rank = MPIR_Process.rank;
-    int in_vci_cs = -1;
     for (int vci = 0; vci < num_vcis; vci++) {
-        MPID_THREAD_CS_ENTER(VCI, MPIDI_VCI_LOCK(vci));
-        in_vci_cs = vci;
-
         MPIDI_OFI_dynamic_process_request_t reqs[2];
         mpi_errno = flush_send(rank, 0, vci, &reqs[0]);
         MPIR_ERR_CHECK(mpi_errno);
@@ -827,15 +822,11 @@ int MPIDI_OFI_flush_send_queue(void)
             mpi_errno = MPIDI_NM_progress(vci, &made_progress);
             MPIR_ERR_CHECK(mpi_errno);
         }
-        MPID_THREAD_CS_EXIT(VCI, MPIDI_VCI_LOCK(vci));
     }
 
   fn_exit:
     return mpi_errno;
   fn_fail:
-    if (in_vci_cs != -1) {
-        MPID_THREAD_CS_EXIT(VCI, MPIDI_VCI_LOCK(in_vci_cs));
-    }
     goto fn_exit;
 }
 
@@ -857,7 +848,11 @@ int MPIDI_OFI_mpi_finalize_hook(void)
     /* Destroy RMA key allocator */
     MPIDI_OFI_mr_key_allocator_destroy();
 
-    if (MPIR_CVAR_NO_COLLECTIVE_FINALIZE) {
+    if (strcmp("sockets", MPIDI_OFI_global.prov_use[0]->fabric_attr->prov_name) == 0) {
+        /* sockets provider need flush any last lightweight send. */
+        mpi_errno = flush_send_queue();
+        MPIR_ERR_CHECK(mpi_errno);
+    } else if (MPIR_CVAR_NO_COLLECTIVE_FINALIZE) {
         /* skip collective work arounds */
     } else if (strcmp("verbs;ofi_rxm", MPIDI_OFI_global.prov_use[0]->fabric_attr->prov_name) == 0
                || strcmp("psm2", MPIDI_OFI_global.prov_use[0]->fabric_attr->prov_name) == 0

--- a/test/mpi/spawn/spawn_rootargs.c
+++ b/test/mpi/spawn/spawn_rootargs.c
@@ -25,8 +25,13 @@ int main(int argc, char *argv[])
 
     if (parent == MPI_COMM_NULL) {
         MPI_Comm_rank(MPI_COMM_WORLD, &rank);
-        MPI_Comm_spawn("./spawn_rootargs", args,        /*MPI_ARGV_NULL, */
-                       5, MPI_INFO_NULL, 0, MPI_COMM_SELF, &child, MPI_ERRCODES_IGNORE);
+        if (rank == 0) {
+            MPI_Comm_spawn("./spawn_rootargs", args,
+                           5, MPI_INFO_NULL, 0, MPI_COMM_WORLD, &child, MPI_ERRCODES_IGNORE);
+        } else {
+            MPI_Comm_spawn(NULL, NULL,
+                           -1, MPI_INFO_NULL, 0, MPI_COMM_WORLD, &child, MPI_ERRCODES_IGNORE);
+        }
         MPI_Barrier(child);
         MPI_Comm_disconnect(&child);
         if (!rank)

